django/db/migrations/state.py

```diff
@@ -94,16 +94,30 @@ def __init__(self, models=None, real_apps=None):
         self.real_apps = real_apps or []
         self.is_delayed = False
         # {remote_model_key: {model_key: [(field_name, field)]}}
-        self.relations = None
+        self.relations = defaultdict(partial(defaultdict, list))
 
     def add_model(self, model_state):
         app_label, model_name = model_state.app_label, model_state.name_lower
         self.models[(app_label, model_name)] = model_state
+        if 'all_relations' in self.__dict__:
```

Comment:
I think changing `if 'all_relations' in self.__dict__:` to `if self.all_relations is None` would make sense .But not using `chached_property` would not cache the relations and perhaps that may increase the workload. 
I am not sure about this.
I think @charettes may have some suggestions for the same.

Reply:
The same amount of caching would be happening in the approach I'm suggesting. It's just that you would be calling `self.resolve_fields_and_relations() / self.all_relations = ...` (e.g. in a method) instead of accessing a cached property. It just seems like the usage in the PR doesn't really match `@cached_property`'s use case. In addition to what I mentioned above, the calls to `self.all_relations` in the PR aren't using the return value, it's just doing that for the caching side effect, which you could do more simply / explicitly.

Reply:
After digging into this, I figured out that there is no need to call `resolve_fields_and_relations` when `relations` are not populated. Or in other words, state alteration methods need not initialize the registry. All they have to do is to alter the `relations` if the `relations` are populated. So would it be fine to just add `if self.relations:` condition in state alteration methods. And only populate the registry when needed.

Reply:
I agree with Chris `cached_property` doesn't fit here. I'd change `relations` to a `property` and update an internal variable `_relation`, e.g.
```python
class ProjectState:
    def __init__(self, models=None, real_apps=None):
        ...
        # {remote_model_key: {model_key: [(field_name, field)]}}
        self._relations = None

    @property
    def relations(self):
        if self._relations is None:
            self.resolve_fields_and_relations()
        return self._relations

    def rename_model(self, app_label, old_name, new_name):
        ...
        # Update changed relations.
        if self._relations is not None:
            old_name_key = (app_label, old_name.lower())
            new_name_key = (app_label, new_name.lower())
            if old_name_key in self._relations:
                self._relations[new_name_key] = self.relations.pop(old_name_key)
            for model_relations in self._relations.values():
                if old_name_key in model_relations:
                    model_relations[new_name_key] = model_relations.pop(old_name_key, '')
        ...
```

Reply:
> Would that be fine? Or is there something I am missing?

`else` is not necessary, there is no need to resolve relations at this point.